### PR TITLE
many: optional recovery keys

### PR DESCRIPTION
**Pull request by bboozzoo**
Optional recovery keys. Based on #11748, with cherry picks from #11757.

The tests that verify idempotency of adding a recovery key will fail until #11800 lands

### CODE REVIEW

File: secboot/encrypt_sb.go [hunk 1]
```diff
@@ -94,17 +94,79 @@ func runSnapFDEKeymgr(args []string, stdin io.Reader) error {
...
+
+func devByPartUUIDFromMount(mp string) (string, error) {
+	partUUID, err := disks.PartitionUUIDFromMountPoint(mp, &disks.Options{
+		IsDecryptedDevice: true,
+	})
+	if err != nil {
+		return "", fmt.Errorf("cannot partition for mount %v: %v", mp, err)
+	}
+	dev := filepath.Join("/dev/disk/by-partuuid", partUUID)
+	logger.Debugf("device by partuuid: %v", dev)
```
Comment (pedronis) [inline secboot/encrypt_sb.go:138]: same comment as in the other PR about this log message

**Review by pedronis — COMMENTED** [review]
did a first pass, thanks for the spread test changes

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -2088,8 +2088,20 @@ func (m *DeviceManager) EnsureRecoveryKeys() (*client.SystemRecoveryKeysResponse
 	if err != nil {
 		return nil, fmt.Errorf("cannot determine ubuntu-data mount point: %v", err)
 	}
-	mountPoints := []string{dataMountPoints[0], boot.InitramfsUbuntuSaveDir}
-	rkey, err := secbootEnsureRecoveryKey(filepath.Join(fdeDir, "recovery.key"), mountPoints)
+	recoveryKeyDevices := []secboot.RecoveryKeyDevice{
+		{
+			Mountpoint: dataMountPoints[0],
+			// TODO ubuntu-data key in install mode?
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go:2094]: A possible solution is that we could explicitly add the key to the keyring at the end of the install. There's some code which does that in `snap-fde-keymgr change-encryption-key`. Then by the time the install device hook runs, it should be possible to generate and add recovery key.

File: overlord/devicestate/devicemgr.go [hunk 3]
```diff
@@ -2088,8 +2088,20 @@ func (m *DeviceManager) EnsureRecoveryKeys() (*client.SystemRecoveryKeysResponse
...
+		},
+		{
+			Mountpoint: boot.InitramfsUbuntuSaveDir,
+			AuthorizationKeyFile: filepath.Join(
+				dirs.SnapFDEDirUnder(filepath.Join(dataMountPoints[0], "system-data")),
+				"ubuntu-save.key",
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go:2100]: This should be correct, the path we expect is `/run/mnt/ubuntu-data/system-data/var/lib/snapd/device/fde`. What I'm now wondering about is that if EnsureRecoveryKeys() is called before /run/mnt/ubuntu-data created HostUbuntuDataForMode will return nil. OTOH this shouldn't happen if the recovery keys can only be added in device install hook during install. Probably warrants a check anyway.

File: overlord/devicestate/devicemgr.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go:2094]: I've updated the comment with more details

File: overlord/devicestate/devicemgr.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go:2100]: Added a check now

File: overlord/devicestate/devicemgr.go [hunk 4]
```diff
@@ -2111,14 +2123,25 @@ func (m *DeviceManager) RemoveRecoveryKeys() error {
 	if err != nil {
 		return fmt.Errorf("cannot determine ubuntu-data mount point: %v", err)
 	}
-	mountPointToKey := make(map[string]string, 2)
+	recoveryKeyDevices := make(map[secboot.RecoveryKeyDevice]string, 2)
 	rkey := filepath.Join(dirs.SnapFDEDir, "recovery.key")
-	mountPointToKey[dataMountPoints[0]] = rkey
+	// TODO figure out appropriate authorization source for each device
+	// depending on mode
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/devicestate/devicemgr.go:2129]: dropped

File: secboot/encrypt.go [hunk 5]
```diff
@@ -27,3 +27,12 @@ const (
...
 	EncryptionTypeDeviceSetupHook EncryptionType = "device-setup-hook"
 )
+
+type RecoveryKeyDevice struct {
+	// Mountpoint of the device
+	Mountpoint string
+	// AuthorizationKeyFile is the path to the key to authorize the
+	// operation, if empty, then it is assumed that the authorization key is
+	// present in the user session keyring
+	AuthorizationKeyFile string
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply secboot/encrypt.go:37]: renamed

**Review by pedronis — APPROVED** [review]
thank you

**Pull request merged**